custom_metrics/wpt_bodies.js

```diff
@@ -0,0 +1,1366 @@
+      try {    
+        let result = {http_header: {values: []}};
+
+        function getHreflangValues(d) {
+          let target = {values: []};
+          let hreflangs = Array.from(d.querySelectorAll('link[rel="alternate"][hreflang]'));
+
+          hreflangs.forEach(e => {
+            target.values.push(e.getAttribute("hreflang"));
+          });
+
+          return target;
+        }
+
+        result.rendered = getHreflangValues(document);
+
+        let rawHtmlDocument = getRawHtmlDocument();
+
+        if (rawHtmlDocument) {
+          result.raw = getHreflangValues(rawHtmlDocument);
+        }
+
+        let linkHeaders = getResponseHeaders("link");
+        if (linkHeaders) {
+          linkHeaders.forEach((h) => {
+            let matches = h.matchAll(/hreflang=['"]?(.*?)['"]/gi);
+
+            for (const match of matches) {
+              let c =match[1];
+              result.http_header.values.push(c);
+            }  
+          })
+        }
+
+        return result; 
+      }
+      catch(e) {
+        return logError("hreflangs", e);
+      }
+    })(),
+
+    // heading information from H1 to H8 
+    // Used by: SEO, Markup
+    // Backup: heading order and length of h1s in almanac.js - not much
+    'headings': (() => { 
+      try {  
+      
+        function processHeading(d, target, n, level) {
+          let html  = n.innerHTML;
+          let text = seoText(n);
+
+          let words = text.match(/\S+/g)?.length;
+
+          if (!target.primary) {
+            if (text.length > 0) { // make this primary
+              let primaryTitle = "";
+              if (d == document) {
+                primaryTitle = renderedPrimaryTitle; 
+              }
+              else {
+                primaryTitle = rawPrimaryTitle; 
+              }
+
+              let snippet = text;
+
+              if (snippet.length > 100) {
+                snippet = snippet.substring(0, 100)+"...";
+              }
+
+              target.primary = {
+                words: words,
+                characters: text.length,
+                matches_title: text.toLowerCase() == primaryTitle?.toLowerCase(),
+                text: snippet,
+                level: level
+              }
+            } else if (html.length > 0) {
+              // looks like a hidden heading
+              target.first_non_empty_heading_hidden = true;
+            }
+          }
+          return {characters: text.length, words: words ?? 0};
+        }
+        
+        function processHeadings(d) {
+          let target = {first_non_empty_heading_hidden: false};
+          for(let l=1; l < 9; l++) {
+            let nodes = Array.from(d.querySelectorAll('h'+l));
+
+            let characters = 0;
+            let words = 0;
+            // if don't have a primary heading yet, search for one.
+            
+            var hs = nodes.map(n => {
+              let h = processHeading(d, target, n, l);
+              characters += h.characters;
```

Comment:
How useful is it to return a sum of all headings for a level combined? You could still derive the average length, but not the distribution/median for example.

Reply:
Yes, there's definitely a lot of "data waste" for metrics that don't show remarkable results. Up to you how to aggregate these.

Reply:
If it's ok with you, can we leave the waste in for this year?

Reply:
SGTM